src/routes/files.ts

```diff
@@ -1,14 +1,412 @@
+
+      if (directoryModification) {
+        if ('doc' in row) {
+          let existingDirectory: StoreDirectoryDocument
+
+          try {
+            existingDirectory = asStoreDirectoryDocument(row.doc)
+          } catch (err) {
+            const apiClientError: ApiClientError = {
+              status: 422,
+              message:
+                `Unable to write files under '${directoryKey}'. ` +
+                `Existing document is not a directory.`
+            }
+            throw apiClientError
+          }
+
+          const directoryDocument: StoreDirectoryDocument = mergeDirectoryFilePointers(
+            existingDirectory,
+            directoryModification
+          )
+
+          // Update document
+          storeDirectoryDocuments.push(directoryDocument)
+        } else {
+          // Insert document
+          storeDirectoryDocuments.push({
+            ...directoryModification,
+            _id: directoryKey
+          })
+        }
+      }
+    })
+  }
+
+  // Prepare Repos Documents:
+
+  const repoKeys = Array.from(repoModifications.keys())
+  const repoFetchResult = await dataStore.fetch({ keys: repoKeys })
+  const storeRepoDocuments: StoreRootDocument[] = []
+
+  // Prepare repo (root) documents
+  repoFetchResult.rows.forEach(row => {
+    const repoKey = row.key
+    const repoModification = repoModifications.get(repoKey)
+
+    if (repoModification) {
+      if ('doc' in row) {
+        let existingRepo: StoreRootDocument
+
+        try {
+          existingRepo = asStoreRootDocument(row.doc)
+        } catch (err) {
+          const apiClientError: ApiClientError = {
+            status: 422,
+            message:
+              `Unable to write files under '${repoKey}'. ` +
+              `Document is not a repo.`
+          }
+          throw apiClientError
+        }
+
+        const repoDocument: StoreRootDocument = mergeDirectoryFilePointers(
+          existingRepo,
+          repoModification
+        )
+
+        storeRepoDocuments.push(repoDocument)
+      } else {
+        // If no existing StoreRootDocument, then we should throw a client error
+        const apiClientError: ApiClientError = {
+          status: 422,
+          message:
+            `Unable to write files under '${repoKey}'. ` +
+            `Document does not exist.`
+        }
+        throw apiClientError
+      }
+    }
+  })
+
+  // Writes:
+
+  // This checks for conflicts and errors in the datastore write results
+  const checkForErrors = (results: DocumentBulkResponse[]) =>
+    results.reduce((success, result) => {
+      if (result.error) {
+        if (result.error === 'conflict') {
+          // For conflict errors, just change the success boolean
+          success = false
+        } else {
+          // For all other errors, throw because it's unexpected
+          throw new Error('Unexpected database error: ' + result.reason)
+        }
+      }
+      return success
+    }, true)
```

Comment:
Move this outside of the request handler. Rename to `checkResultsForRevMismatch`.

Reply:
Resolution: https://github.com/EdgeApp/edge-sync-server/commit/e41d00252299e57f88a240b02f1eb37dbaaf2842#diff-2196a720a13a3fd21839e328edb702c97e36909db9cc12f2175981232fdc3ce7R342